_posts/2021-06-22-type-phantoms-and-safe-apis.md

```diff
@@ -0,0 +1,712 @@
+
+    res.send(await handler({ path: req.params, body: req.body }, context))
+  }
+```
+
+If the *type* of `Context` also depends on the endpoint, the utilities would need changes as well.
+
+## Further improvements
+
+That's quite a few changes and tweaks already! Turns out it's not easy to write glue code and
+utilities that would be usable out of the box. Project requirements and techniques vary, and the
+glue code needs adjustments as well.
+
+We've not even touched on many common API features that may be needed:
+
+- How is authentication handled? What about authorization?
+- What about specific request and response headers?
+- How to make multiple separate APIs type safe?
+- Can we make route handling even more safe, for example with path uniqueness checks?
+
+Some of these are straightforward using techniques we already explored. That means even more changes
+to the common utilities though. Could we develop the core functionality further to make it more
+generic and support various kinds of API structures?
+
+### Adapting to different API structures
+
+Using custom adapters would help with some of these issues. The generic utilities are less useful
+then, but the request and response conversions can be customized. The client-side implementation
+could look like this:
+
+```typescript
+// lib/Client.ts
+
+type MakeRequest<EP extends Endpoint = Endpoint> =
+  (params: Omit<EP, 'result'>) => Promise<EP['result']>
+
+export function createHandlers<T extends Endpoints>(
+  makeRequest: MakeRequest,
+  routes: Routes<T>,
+): Handlers<T> {
+  /* ... */
+}
+```
+
+Now the common utility functions don't know the details of how to make requests, and those details
+can be tweaked freely. Some features still need type changes. For example, using the `Result` type
+means changing the `MakeRequest` type as well.
+
+The server implementation could use the same principles to extract library logic and enable
+customization.
+
+### Simplifying types
+
+Making things simpler can make them easier to use as well. The utility types are quite complex, and
+perhaps some could even be removed entirely? It's easy to make some types simpler, but type
+inference can suffer, and the error messages could become more cryptic.
+
+It's also possible to derive all the API model types from the server-side implementation. Then
+separate types wouldn't have to be written at all. However, this would make the API model less
+understandable in most cases and make some error messages incomprehensible.
+
+Despite these challenges, there's probably a lot that could be improved, without adverse effects on
+readability.
+
+### Stop being so RESTful?
+
+If you are just writing function calls, do you need separate body, path and query parameters? Are
+the different HTTP methods useful?
+
+If it's an external API, sticking to typical REST conventions will be more familiar to the API
+users. For an internal API, the API structure may not matter that much when it's abstracted away.
```

Comment:
Interesting points that ring true. Im a little curious: What kinds of external APIs would this approach work well with? Maybe ones that also have internal clients or maybe even some SDK?

The way I've viewed this so far has been along the lines of "full-stack React / framework X". That seems to be the value-proposition of Blitz _(although Next itself aligns with serverless)_, and a major use-case for tRPC.

Reply:
Yeah, mostly referring to an API that's *also* used externally. In that case these tweaks may not be a good idea. Could always provide that SDK of course.